lib/Widgets/MessageDialog.vala

```diff
@@ -0,0 +1,242 @@
+        }
+        
+        set {
+            image.set_from_gicon (value, Gtk.IconSize.DIALOG);
+        }
+    }
+
+    /**
+     * The {@link Gtk.Label} that displays the {@link Granite.MessageDialog.primary_text}.
+     * 
+     * Most of the times, you will only want to modify the {@link Granite.MessageDialog.primary_text} string, 
+     * this is available to set additional properites like {@link Gtk.Label.use_markup} if you wish to do so.
+     */
+    public Gtk.Label primary_label { get; construct; }
+
+    /**
+     * The {@link Gtk.Label} that displays the {@link Granite.MessageDialog.secondary_text}.
+     *
+     * Most of the times, you will only want to modify the {@link Granite.MessageDialog.secondary_text} string, 
+     * this is available to set additional properites like {@link Gtk.Label.use_markup} if you wish to do so.
+     */
+    public Gtk.Label secondary_label { get; construct; }
+
+    /**
+     * The {@link Gtk.ButtonsType} value to display a set of buttons
+     * in the dialog.
+     * 
+     * By design, some actions are not acceptable and such action values will not be added to the dialog, these include:
+     *
+     *  * {@link Gtk.ButtonsType.OK}
+     *  * {@link Gtk.ButtonsType.YES_NO}
+     *  * {@link Gtk.ButtonsType.OK_CANCEL}
+     *
+     * If you wish to provide more specific actions for your dialog
+     * pass a {@link Gtk.ButtonsType.NONE} to {@link Granite.MessageDialog} and manually
+     * add those actions with {@link Gtk.Dialog.add_buttons} or {@link Gtk.Dialog.add_action_widget}.
+     */
+    public Gtk.ButtonsType buttons { 
+        construct {
+            switch (value) {
+                case Gtk.ButtonsType.NONE:
+                    break;
+                case Gtk.ButtonsType.CLOSE:
+                    add_button (_("_Close"), Gtk.ResponseType.CLOSE);
+                    break;
+                case Gtk.ButtonsType.CANCEL:
+                    add_button (_("_Cancel"), Gtk.ResponseType.CANCEL);
+                    break;
+                case Gtk.ButtonsType.OK:
+                case Gtk.ButtonsType.YES_NO:
+                case Gtk.ButtonsType.OK_CANCEL:
+                    warning ("Unsupported GtkButtonsType value");
+                    break;
+                default:
+                    warning ("Unknown GtkButtonsType value");
+                    break;
+            }
+        }
+    }
+
+    /**
+     * The message grid that contains all the message widgets: the {@link Granite.MessageDialog.primary_label},
+     * {@link Granite.MessageDialog.secondary_label} and the image widget.
+     * 
+     * This grid can be used to add any custom widget to the message area such as a {@link Gtk.ComboBox} or {@link Gtk.CheckButton}.
+     * Note that after adding such widget you will need to call {@link Gtk.Widget.show} or {@link Gtk.Widget.show_all} on the widget itself for it to appear in the dialog.
+     */
+    public Gtk.Grid message_grid { get; construct; }
```

Comment:
I'm not a huge fan of exposing the grid that contains the actual message labels for a couple reasons:
* It seems pretty non-intuitive that I have to start packing my widgets at `[1, 2]`. This requires me to know about the internal structure of the dialog and if I don't, I can attach widgets overlapping with the pre-populated ones or at the incorrect alignment.

* additional widgets packed into the dialog should be separated from the secondary text label by 24px (effectively, we double space between sections). It would be nice if this was done automatically for me instead of requiring me to know this.

I think it would be more intuitive to use a grid `content_area` and override Gtk.Dialog's `get_content_area` if possible. But if not, I think the publicly exposed grid should not include the icon and message text.

Reply:
@danrabbit it is not possible to override the `get_content_area` of the `Gtk.Dialog` class. We have 2 main solutions to this:

* Add a new special container inside the `message_grid` and expose it through the API.
* Keep the `message_grid` property and describe in the documentation it's details and structure including all the values like `[1, 2]`.

The first approach is what `Gtk.MessageDialog` uses: https://valadoc.org/gtk+-3.0/Gtk.MessageDialog.get_message_area.html, but with a difference that it's returning a `Gtk.Box` instead of a grid.

Reply:
I think adding another container would be much better. I don't think it feels good to expect developers to know about the internal structure of the widget